src/patterns.jl

```diff
@@ -0,0 +1,63 @@
+
+"""
+    Pattern
+
+A pattern is just a wrapper around a `Tuple{Vararg{Symbol}}` which enables searching for
+matching components and dimension paths in a Dataset.
+Special symbols `:_` and `:__` are used as wildcards, similar to `*` and `**` in glob
+pattern matching.
+"""
+struct Pattern
+    segments::Tuple{Vararg{Symbol}}
+end
+
+Pattern(segments::Symbol...) = Pattern(segments)
+Base.convert(::Type{Pattern}, x::Tuple{Vararg{Symbol}}) = Pattern(x)
+
+# Primary functionality is to support identify tuples that match the constraint
+# NOTE: I'm not entirely sure if we should overload `in` for this algorithm
+function Base.in(item::Tuple{Vararg{Symbol}}, pattern::Pattern)
+    pat = pattern.segments
+    i = 1   # item  index
+    j = 1   # contraint index
+    n = length(item)
+    m = length(pat)
```

Comment:
the algorithm might be easier to understand if we name these variables as `num_items` and `num_pats`?

Reply:
Hmm, I feel like `i`/`j`and `n`/`m` are pretty common variable names in iteration algorithms. Though I wonder if this would be more readable if it was written to use `iterate` instead?

Reply:
Alright, function rewrite to use `iterate`.

Reply:
yeah that change actually made it much easier to follow